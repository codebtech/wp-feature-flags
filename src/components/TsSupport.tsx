import { useMemo } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { Flag } from '../../types';

import Snippet from './Snippet';

const TsSupport = ({ flag }: { flag: Flag }): JSX.Element => {
	const tsSnippet = useMemo(() => {
		return `
declare namespace mrFeatureFlags {
    export interface FeatureFlagProps {
        isEnabled: (flag: string) => boolean;
    }
}
declare global {
    interface Window {
        mrFeatureFlags: mrFeatureFlags.FeatureFlagProps;
    }
}
export {};`;
	}, []);

	const tsIgnoreSnippet = useMemo(() => {
		return `import domReady from '@wordpress/dom-ready';
        domReady(function () {
            //@ts-ignore
            if (window.mrFeatureFlags.isEnabled('${flag.name}')) {
                // js code goes here...
            }
         });`;
	}, [flag.name]);

	return (
		<div className="mr-feature-flag-php-snippet-container">
			<h3>{__('TypeScript support', 'mr-feature-flags')}</h3>
			<p>
				Create `flags.d.ts` file at the root of the plugin / theme to
				add typescript declaration for feature flags.
			</p>
			<Snippet data={tsSnippet} language={'typescript'} />
			<p>
				Alternatively, you can ignore typescript warning using below
				approach
			</p>
			<Snippet data={tsIgnoreSnippet} language={'typescript'} />
		</div>
	);
};

export default TsSupport;
