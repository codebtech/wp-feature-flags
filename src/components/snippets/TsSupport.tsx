import { useMemo } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import Clipboard from '../common/Clipboard';

import Snippet from './Snippet';

const TsSupport = (): JSX.Element => {
	const tsSnippet = useMemo(() => {
		return `declare namespace mrFeatureFlags {
	export interface FeatureFlagProps {
        isEnabled: (flag: string) => boolean;
    }
}
declare global {
    interface Window {
        mrFeatureFlags: mrFeatureFlags.FeatureFlagProps;
    }
}
export {};`;
	}, []);

	return (
		<div className="codeb-feature-flags-ts-snippet-container">
			<h3>{__('TypeScript support', 'codeb-feature-flags')}</h3>
			<p>
				Create a file named{' '}
				<span className="codeb-feature-flags-filename">flags.d.ts</span>{' '}
				at the entrypoint of TypeScript code for the plugin/theme (
				usually
				<span className="codeb-feature-flags-slug"> src</span> directory
				) and add the following declaration.
			</p>
			<Clipboard text={tsSnippet} />
			<Snippet data={tsSnippet} language={'typescript'} />
		</div>
	);
};

export default TsSupport;
