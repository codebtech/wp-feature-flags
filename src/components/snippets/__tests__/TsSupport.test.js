import { render } from '@testing-library/react';
import TsSupport from '../TsSupport';

describe('Typescript snippet component', () => {
	test('renders without any error', async () => {
		const { asFragment } = render(<TsSupport />);
		expect(asFragment()).toMatchSnapshot();
	});
});
