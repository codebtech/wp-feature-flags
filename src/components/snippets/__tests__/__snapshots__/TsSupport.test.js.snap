// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript snippet component renders without any error 1`] = `
<DocumentFragment>
  <div
    class="codeb-feature-flags-ts-snippet-container"
  >
    <h3>
      TypeScript support
    </h3>
    <p>
      Create a file named 
      <span
        class="codeb-feature-flags-filename"
      >
        flags.d.ts
      </span>
       at the entrypoint of TypeScript code for the plugin/theme ( usually
      <span
        class="codeb-feature-flags-slug"
      >
         src
      </span>
       directory ) and add the following declaration.
    </p>
    <button
      aria-label="Copy to clipboard"
      aria-pressed="false"
      class="components-button codeb-feature-flags-clipboard-base is-tertiary has-icon"
      type="button"
    >
      <span
        class="dashicon dashicons dashicons-clipboard"
      />
    </button>
    <pre
      style="display: block; overflow-x: auto; background: rgb(43, 43, 43); color: rgb(248, 248, 242); padding: 0.5em;"
    >
      <code
        class="language-typescript"
        style="white-space: pre;"
      >
        <span
          style="color: rgb(220, 198, 224);"
        >
          declare
        </span>
        <span>
           
        </span>
        <span
          style="color: rgb(220, 198, 224);"
        >
          namespace
        </span>
        <span>
           codebFeatureFlags {

        </span>
        <span>
          	
        </span>
        <span
          style="color: rgb(220, 198, 224);"
        >
          export
        </span>
        <span>
           
        </span>
        <span
          style="color: rgb(220, 198, 224);"
        >
          interface
        </span>
        <span>
           FeatureFlagProps {

        </span>
        <span>
                  
        </span>
        <span
          class="hljs-attr"
        >
          isEnabled
        </span>
        <span>
          : 
        </span>
        <span
          class="hljs-function"
        >
          (
        </span>
        <span
          class="hljs-function"
          style="color: rgb(245, 171, 53);"
        >
          flag: 
        </span>
        <span
          class="hljs-function"
          style="color: rgb(245, 171, 53);"
        >
          string
        </span>
        <span
          class="hljs-function"
        >
          ) =&gt;
        </span>
        <span>
           
        </span>
        <span
          style="color: rgb(245, 171, 53);"
        >
          boolean
        </span>
        <span>
          ;

        </span>
            }
}

        <span />
        <span
          style="color: rgb(220, 198, 224);"
        >
          declare
        </span>
        <span>
           
        </span>
        <span
          style="color: rgb(245, 171, 53);"
        >
          global
        </span>
        <span>
           {

        </span>
        <span>
              
        </span>
        <span
          style="color: rgb(220, 198, 224);"
        >
          interface
        </span>
        <span>
           Window {

        </span>
        <span>
                  
        </span>
        <span
          class="hljs-attr"
        >
          codebFeatureFlags
        </span>
        <span>
          : codebFeatureFlags.FeatureFlagProps;

        </span>
            }
}

        <span />
        <span
          style="color: rgb(220, 198, 224);"
        >
          export
        </span>
        <span>
           {};
        </span>
      </code>
    </pre>
  </div>
</DocumentFragment>
`;
